src/Data/Constraint/Nat.hs

```diff
@@ -69,11 +78,18 @@ type Divides n m = n ~ Gcd n m
 newtype Magic n = Magic (KnownNat n => Dict (KnownNat n))
 #endif
 
-magic :: forall n m o. (Integer -> Integer -> Integer) -> (KnownNat n, KnownNat m) :- KnownNat o
+magicNNN :: forall n m o. (Nat.Natural -> Nat.Natural -> Nat.Natural) -> (KnownNat n, KnownNat m) :- KnownNat o
+#if MIN_VERSION_base(4,18,0)
+magicNNN f = Sub $ withKnownNat @o (unsafeSNat (natVal (Proxy @n) `f` natVal (Proxy @m))) Dict
+#else
+magicNNN f = Sub $ unsafeCoerce (Magic Dict) (natVal (Proxy :: Proxy n) `f` natVal (Proxy :: Proxy m))
```

Comment:
Would it make sense to use `TypeApplications` in both branches now?

I wonder how the `#else` branch worked previously, as it looks like `unsafeCoerce` were coercing an `Integer`, but `SNat`  has backed by `Natural`; but the new code is all `Natural` so should be right.

Reply:
> Would it make sense to use `TypeApplications` in both branches now?

Yes, a fair point. I've pushed an additional commit which uses `TypeApplications` in a more uniform manner.

> I wonder how the `#else` branch worked previously, as it looks like `unsafeCoerce` were coercing an `Integer`, but `SNat` has backed by `Natural`; but the new code is all `Natural` so should be right.

Yes, that's something which has bothered me as well, as the previous code seemed like it was unsound. I spent some time trying to come up with a test case that demonstrated the potential for bad runtime behavior, but I could never manage to do so. Perhaps it worked because the memory layouts for `Integer` and `Natural` are so similar?

In any case, now that we use `Natural` here, this should no longer be a concern.